### PR TITLE
minicpm4 support

### PR DESCRIPTION
**Pull request by fjaijf**
#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### Description

模型前20个token前向精度对齐，log_diff为0.029
模型前10个token生成一致
模型训练部分loss为
[2026-05-15 10:51:23,095] [    INFO] - loss: 2.41923141, learning_rate: 9.18e-06, global_step: 41, current_memory_allocated: 7.4686994552612305, current_memory_reserved: 8.304634094238281, max_memory_allocated: 7.808573484420776, max_memory_reserved: 8.304634094238281, interval_runtime: 32.228, interval_samples_per_second: 0.031, interval_steps_per_second: 0.031, ppl: 11.237219180656735, progress_or_epoch: 0.082, cpu_used_memory: 30.35, cpu_available_memory: 158.19
[2026-05-15 10:51:23,095] [    INFO] - [DataLoad global_step: 41] data_load_time: 0.27 ms (accumulated over 1 micro-batches)
[2026-05-15 10:51:23,645] [    INFO] - loss: 1.80299592, learning_rate: 9.16e-06, global_step: 42, current_memory_allocated: 7.4686994552612305, current_memory_reserved: 8.304634094238281, max_memory_allocated: 7.808573484420776, max_memory_reserved: 8.304634094238281, interval_runtime: 0.5496, interval_samples_per_second: 1.8195, interval_steps_per_second: 1.8195, ppl: 6.0677989007936, progress_or_epoch: 0.084, cpu_used_memory: 30.35, cpu_available_memory: 158.19
[2026-05-15 10:51:23,645] [    INFO] - [DataLoad global_step: 42] data_load_time: 0.08 ms (accumulated over 1 micro-batches)
[2026-05-15 10:51:24,198] [    INFO] - loss: 2.11331582, learning_rate: 9.14e-06, global_step: 43, current_memory_allocated: 7.4686994552612305, current_memory_reserved: 8.304634094238281, max_memory_allocated: 7.808573484420776, max_memory_reserved: 8.304634094238281, interval_runtime: 0.5537, interval_samples_per_second: 1.806, interval_steps_per_second: 1.806, ppl: 8.275636361515012, progress_or_epoch: 0.086, cpu_used_memory: 30.35, cpu_available_memory: 158.19
[2026-05-15 10:51:24,199] [    INFO] - [DataLoad global_step: 43] data_load_time: 0.08 ms (accumulated over 1 micro-batches)
[2026-05-15 10:51:24,762] [    INFO] - loss: 1.32196712, learning_rate: 9.12e-06, global_step: 44, current_memory_allocated: 7.4686994552612305, current_memory_reserved: 8.304634094238281, max_memory_allocated: 7.808573484420776, max_memory_reserved: 8.304634094238281, interval_runtime: 0.5634, interval_samples_per_second: 1.7751, interval_steps_per_second: 1.7751, ppl: 3.7507923837707833, progress_or_epoch: 0.088, cpu_used_memory: 30.37, cpu_available_memory: 158.17
[2026-05-15 10:51:24,762] [    INFO] - [DataLoad global_step: 44] data_load_time: 0.08 ms (accumulated over 1 micro-batches)
[2026-05-15 10:51:25,315] [    INFO] - loss: 1.83120203, learning_rate: 9.1e-06, global_step: 45, current_memory_allocated: 7.4686994552612305, current_memory_reserved: 8.304634094238281, max_memory_allocated: 7.808573484420776, max_memory_reserved: 8.304634094238281, interval_runtime: 0.5529, interval_samples_per_second: 1.8086, interval_steps_per_second: 1.8086, ppl: 6.241384482706879, progress_or_epoch: 0.09, cpu_used_memory: 30.36, cpu_available_memory: 158.18

### CODE REVIEW

File: tests/transformers/minicpm4/test_modeling_minicpm4.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+    random_attention_mask,
+)
+
+
+class MiniCPMModelTester:
+    def __init__(
+        self,
+        parent,
+        vocab_size=32000,
+        hidden_size=1024,
```
Comment (liuhao2638) [inline tests/transformers/minicpm4/test_modeling_minicpm4.py]: P1 **优先级：P1**

这个测试默认配置会实例化接近真实规模的模型：`hidden_size=1024`、`num_hidden_layers=32`，同时 `MiniCPMConfig` 默认 `intermediate_size=11008`、`vocab_size=32000`。`ModelTesterMixin` 会反复构造模型，这会让单测非常慢甚至在普通 CI/开发机上 OOM，导致新增覆盖不可用。请改成 micro 配置，并把 `intermediate_size` 也显式传入 `MiniCPMConfig`。

建议形状：

```python
vocab_size=256,
hidden_size=64,
intermediate_size=128,
num_hidden_layers=2,
num_attention_heads=4,
num_key_value_heads=2,
...
self.intermediate_size = intermediate_size
...
intermediate_size=self.intermediate_size,
```

File: tests/transformers/minicpm4/test_modeling_minicpm4.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        )
+        if self.parent.use_labels:
+            self.parent.assertTrue((result_position_id[1] == result_no_position_id[1]).all())
+        else:
+            self.parent.assertTrue((result_position_id[0] == result_no_position_id[0]).all())
+
+    def create_and_check_gqa_model(self, config, input_ids, input_mask, *args):
+        model = MiniCPMForCausalLM(config)
+        config.num_key_value_heads = 8  # gqa
+        config.apply_rope_fusion = True
```
Comment (liuhao2638) [inline tests/transformers/minicpm4/test_modeling_minicpm4.py]: P2 **优先级：P2**

这里先构造了 `MiniCPMForCausalLM(config)`，再把 `config.num_key_value_heads` 改成 8；注意力层在 `__init__` 里已经把 `num_key_value_heads` 缓存到模块属性里了，所以这个测试实际上没有验证 8 个 KV head 的 GQA 路径。请在构造模型前修改配置，或者单独创建一个 GQA config。

```suggestion
        config.num_key_value_heads = 8  # gqa
        config.apply_rope_fusion = True
        model = MiniCPMForCausalLM(config)
```

File: paddleformers/transformers/minicpm4/modeling.py [hunk 3]
```diff
@@ -0,0 +1,2069 @@
...
+    if head_mask is not None:
+        head_mask = _convert_head_mask_to_5d(head_mask, num_hidden_layers)
+        if is_attention_chunked is True:
+            head_mask = head_mask.unsqueeze(-1)
+    else:
+        head_mask = [None] * num_hidden_layers
+    return head_mask
+
+
+setattr(paddleformers.transformers.model_utils.PretrainedModel, "get_head_mask", _get_head_mask)
```
Comment (liuhao2638) [inline paddleformers/transformers/minicpm4/modeling.py]: P1 **优先级：P1**

这里在导入 `minicpm4.modeling` 时直接给全局 `PretrainedModel` 注入 `get_head_mask`、`device` 和 `post_init`。这会影响同一进程里所有已导入或后续导入的模型；尤其 `post_init` 会改变其他模型里 `self.post_init()` 的解析结果，属于 MiniCPM 之外的全局副作用。请把这些 helper 收敛到 `MiniCPMPreTrainedModel`（或直接使用现有 `init_weights`/metaclass 初始化流程），不要修改共享基类。

可以按这个形状局部化：

```python
class MiniCPMPreTrainedModel(PretrainedModel):
    ...

    def get_head_mask(self, head_mask, num_hidden_layers, is_attention_chunked=False):
        ...

    def post_init(self):
        self.init_weights()
```

**Review by liuhao2638 — CHANGES_REQUESTED** [review]
已完成首轮 review。当前有需要修复后再合入的行级问题，细节已放在 inline review comments；另外也标出了一处测试有效性的中优先级问题。CI 当前显示通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: tests/transformers/minicpm4/test_modeling_minicpm4.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+            labels=input_ids if self.parent.use_labels else None,
+            return_dict=self.parent.return_dict,
+        )
+        if self.parent.use_labels:
+            self.parent.assertTrue((result_position_id[1] == result_no_position_id[1]).all())
+        else:
+            self.parent.assertTrue((result_position_id[0] == result_no_position_id[0]).all())
+
+    def create_and_check_gqa_model(self, config, input_ids, input_mask, *args):
+        config.num_key_value_heads = 8  # gqa
```
Comment (risemeup1111) [inline tests/transformers/minicpm4/test_modeling_minicpm4.py]: P1 **优先级：P1**

这里当前配置已经把 `num_attention_heads` 缩到 4，但又把 `num_key_value_heads` 改成 8。`MiniCPMAttention.__init__` 会计算 `num_key_value_groups = self.num_heads // self.num_key_value_heads`，因此这里变成 `4 // 8 == 0`；eager 路径会把 KV head repeat 成 0 个，SDPA/GQA 也不是合法的 head 数组合，导致这个单测路径会失败而不是验证 GQA。请改成能整除 query heads 的 KV head 数，或同步调整 attention heads。

```suggestion
        config.num_key_value_heads = 2  # gqa, must divide num_attention_heads=4
```

File: tests/transformers/minicpm4/test_modeling_minicpm4.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+class MiniCPMModelTester:
+    def __init__(
+        self,
+        parent,
+        vocab_size=256,
+        hidden_size=64,
+        num_hidden_layers=2,
+        num_attention_heads=4,
+        num_key_value_heads=2,
```
Comment (risemeup1111) [inline tests/transformers/minicpm4/test_modeling_minicpm4.py:51]: P1 **优先级：P1**

这次已经把 `hidden_size/layers/vocab` 缩成 micro 配置，但 `get_config()` 仍没有传 `intermediate_size`，所以 `MiniCPMConfig` 会继续使用默认 `intermediate_size=11008`。每层 FFN 仍会创建 `64 x 11008` 级别的矩阵，`ModelTesterMixin` 多次实例化时还是会让单测明显变慢/占用过高。请把 tester 的 `intermediate_size` 一起设小，并在 `get_config()` 传入。

```suggestion
        vocab_size=256,
        hidden_size=64,
        intermediate_size=128,
        num_hidden_layers=2,
        num_attention_heads=4,
```

同时在 `__init__` 中保存并传入配置：

```python
self.intermediate_size = intermediate_size
...
intermediate_size=self.intermediate_size,
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查新提交。之前关于全局修改 `PretrainedModel` 的问题已经移除，GQA 构造顺序也已调整；但测试配置里仍有会影响单测可用性的阻塞问题，细节已放在 inline review comments。CI 当前仍在运行，建议修复后结合 CI 结果再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
已复查最新提交，之前两个行级问题（GQA 头数配置和测试用例 `intermediate_size` 配置）都已修复。本轮没有发现新的阻塞代码问题。

当前 Lint 和基础检查已通过，GPU/XPU/集成测试仍在运行，建议等待这些 CI 结果完成后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/minicpm4/configuration.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    >>> configuration = MiniCPMConfig()
+
+    >>> # Initializing a model from the minicpm-7b style configuration
+    >>> model = MiniCPMModel(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
+    model_type = "minicpm"
```
Comment (risemeup1111) [inline paddleformers/transformers/minicpm4/configuration.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

最新合入 `develop` 后仓库里已经存在 `paddleformers.transformers.minicpm`，它同样导出 `MiniCPMConfig/MiniCPMModel/MiniCPMForCausalLM` 且使用 `model_type="minicpm"`。这里 MiniCPM4 继续复用同一个 `model_type` 和同名顶层符号，会让 Auto/Lazy 映射无法唯一定位 MiniCPM4：`paddleformers/transformers/__init__.py` 中后面的 `minicpm.*` 会覆盖前面的 `minicpm4.*` 同名符号，`CONFIG_MAPPING_NAMES`/`MODEL_NAMES_MAPPING` 里也出现重复的 `minicpm` 键。结果是 `tests/transformers/minicpm4/test_modeling_minicpm4.py` 的顶层导入可能测到已有 `minicpm`，而不是本 PR 的 `minicpm4` 实现；用户通过 `AutoConfig/AutoModel` 加载时也可能解析到错误模块或找不到类。

请把 MiniCPM4 的模型类型和公开符号与已有 MiniCPM 解耦，或明确迁移并删除旧映射；测试也应显式导入被测模块。建议形状：

```python
class MiniCPM4Config(PretrainedConfig):
    model_type = "minicpm4"

CONFIG_MAPPING_NAMES = OrderedDict([
    ...
    ("minicpm4", "MiniCPM4Config"),
])
MODEL_NAMES_MAPPING = OrderedDict([
    ...
    ("minicpm4", "MiniCPM4Model"),
])
# tests:
from paddleformers.transformers.minicpm4 import MiniCPM4Config, MiniCPM4ForCausalLM, MiniCPM4Model
```

如果必须兼容外部 checkpoint 里的 `"model_type": "minicpm"`，也需要只保留一个 `minicpm` owner，并通过 `architectures`/兼容迁移逻辑明确路由到 MiniCPM4，不能同时保留两个同名映射。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。之前由 `risemeup1111` 提出的两个 P1 测试配置问题仍然看起来已修复；不过这次合入后与已有 `minicpm` 模块产生了新的命名/Auto 映射冲突，属于需要先修复的阻塞问题，细节已放在 inline review comment。

当前还有部分 CI 仍在排队或运行，建议修复后结合完整 CI 结果再继续推进。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/minicpm4/configuration.py (hunk 6 above)
Comment (risemeup1111) [reply paddleformers/transformers/minicpm4/configuration.py]: 新提交已完成 `MiniCPM4Config`/模型类改名，并将 `model_type` 改为 `minicpm4`；但两张 Auto 注册表仍未同步：当前新增的 `CONFIG_MAPPING_NAMES` 和 `MODEL_NAMES_MAPPING` 行仍分别重复注册 `("minicpm", "MiniCPMConfig")`、`("minicpm", "MiniCPMModel")`，没有注册 `minicpm4`。因此新加的 `test_auto_mappings` 在 `AutoConfig.from_pretrained()` 处仍无法解析该配置。请按原建议改为 `("minicpm4", "MiniCPM4Config")` 和 `("minicpm4", "MiniCPM4Model")`，并提交新的 commit。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [GQA 头数配置](https://github.com/PaddlePaddle/PaddleFormers/pull/4466#discussion_r3524544474) | P1 | ✅ |
| 2 | [测试模型规模配置](https://github.com/PaddlePaddle/PaddleFormers/pull/4466#discussion_r3524544902) | P1 | ✅ |
| 3 | [MiniCPM4 Auto 映射](https://github.com/PaddlePaddle/PaddleFormers/pull/4466#discussion_r3612220813) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/minicpm4/configuration.py (hunk 6 above)
Comment (risemeup1111) [reply paddleformers/transformers/minicpm4/configuration.py]: 已验证提交 `33e6f74c`：两张 Auto 注册表已分别改为 `("minicpm4", "MiniCPM4Config")` 和 `("minicpm4", "MiniCPM4Model")`，同时保留了已有 `minicpm` 注册；该问题已修复。

**Review by risemeup1111 — APPROVED** [review]

File: paddleformers/transformers/minicpm4/configuration.py (hunk 6 above)
Comment (risemeup1111) [reply paddleformers/transformers/minicpm4/configuration.py]: 在当前头 `0f5e57dc` 复核发现该问题再次出现：`paddleformers/transformers/auto/configuration.py` 的 `MODEL_NAMES_MAPPING` 先注册了 `("minicpm", "MiniCPMModel")`，后面又注册 `("minicpm", "MiniCPM")`；`OrderedDict` 最终保留后者，而 `paddleformers/transformers/minicpm/modeling.py` 只导出 `MiniCPMModel`，不存在 `MiniCPM`。因此通过 `MODEL_MAPPING[MiniCPMConfig]` 的惰性加载会找不到模型类。请删除重复注册并保留唯一正确的 MiniCPM 映射，同时确保 MiniCPM4 继续使用 `minicpm4`。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/minicpm4/configuration.py (hunk 6 above)
Comment (risemeup1111) [reply paddleformers/transformers/minicpm4/configuration.py]: 已验证提交 `e5b3ded3`：重复的 `("minicpm", "MiniCPM")` 已删除，当前两组注册分别唯一指向 `MiniCPMConfig`/`MiniCPMModel` 和 `MiniCPM4Config`/`MiniCPM4Model`；新增测试也同时断言了两条 `MODEL_MAPPING` 路径。该问题已修复。

**Review by risemeup1111 — APPROVED** [review]
